### DISCUSSION TITLE
Right way to write transient data

### DISCUSSION THREAD

**Discussion by flotus1** [CGNS/CGNS#297] [CGNStalk]
Loosely based on the example given in the documentation, I tried writing my own code that produces a CGNS file with solution data for several time steps. One of the requirements I have is this: I want the CGNS file to be valid and readable after each time step is done. So unlike the example, which only writes some of the metadata at the very end.
I have come up with a few implementations for this, but so far all of them are facing the same issue: performance. It takes an increasingly higher amount of time to write each subsequent time step to the file. See results here
![cgns_timing](https://user-images.githubusercontent.com/57221355/131976515-550d6395-15f4-4b3b-a347-ee6228cec632.png)
The particular example that got these results wrote all solution nodes and metadata at the beginning, and then only added the solution data to each node one by one. I changed the order of operations around, but to no avail. Also replicating the order of operations in the documentation (write mesh, generate solution nodes, write data to solution nodes, write metadata at the very end) exhibits similar performance issues.
What's even worse: the more time steps I intend to add to the file, the longer it takes already for the initial time step. What is 0.2-0.3s in this example becomes several seconds for more than 1000 time steps.

Now the only way around this I found so far was to never close the CGNS file at any point other than at the very end. Which leaves a lot of data in flight, and leads to an unreadable/corrupted CGNS file if I abort the program too soon.

I'm on Linux, using CGNS Version 4.2 along with HDF5. Programming in Fortran, and the disk this was tested on is an NVMe-SSD capable of much higher write speeds than I am getting here.
While this code is running (so far only single-core) the load for this core is reported at 100%, with very little write/read activity on the disk.
Is this kind of behavior expected? What's the proper way to get a readable CGNS file after each time step? Is there a way to "flush" the data to disk, without causing the overhead I measured?

**Comment by scspiegel** [comment]
I saw a similar thing when I was adding transient solution output to my code (Modern Fortran, Linux, Parallel CGNS). There might be a better way to do this, but this is the method that I found to work quite well:

1. Before the solver starts, write out file "solution.grid.cgns" which contains all the grid data, e.g., grid coordinates, grid connectivity, etc.
2. When you want to output the solution at a given time, write out the file "solution.#.cgns" where "#" can be whatever you want (I use the time step #). In this file, you only write out the solution data, and create a CGNS link node to the grid file. Somewhere in the process of writing the solution file, store the current time value in an array that gives the time for each solution file that has been written.
3. At the end of the simulation, write out the file "solution.time.cgns". This file contains a CGNS link node to the grid file and every solution file that was written. Here, you output the solution time array into this file so that the corresponding time value is assigned to each solution file.

This has one major disadvantage: it can create a bunch of output files in your output directory depending on the frequency you output solution files. However, this can actually be an advantage when your grid/solution files are many GBs.

The advantages though:
1. It keeps the storage of all the solution and grid data at an absolute minimum (minimal repeat data across files), as if you had created a single file with the same data.
2. You can access each file separately if needed. You can open up the grid file if you want to examine just the grid. You can open any individual solution file to examine the solution at just that point in time. You can open up multiple solution files if you want to look at the solution at a different frequency than your output frequency. Finally, you can open up the time file to look at the full transient solution which is what you are trying to achieve here.
3. I haven't had any issues with slowdowns related to writing transient data.
4. All CGNS files are closed immediately after writing so you don't have to worry about corrupt CGNS files. This also means you can examine solution files while the simulation is still running.

Hopefully this helps, or at least gives you an idea for a way to get around the slowdowns you are experiencing.

**Reply by brtnfld** [reply]
I think @scspiegel 's method is the way to go. 

I'm not sure if the time increase is in HDF5 or CGNS. 
I would be curious if this would help (with CGNS develop), assuming the slowdown is in the HDF5 layer: 

```
 INTEGER, TARGET :: value_f
 value_f = 16*1024*1024
 CALL cg_configure_f(CG_CONFIG_HDF5_MD_BLOCK_SIZE, C_LOC(value_f), ier)
```
Add this before creating the file.

There was a flush API but it was removed. If the slowdown is in the CGNS layer that would not help anyway. It would help the corrupt file issue, but it would not improve the metadata issue during write if that is causing the slowdown.

**Reply by flotus1** [reply]
Thanks for your feedback, it really helps a lot. That's something I probably would not have thought of myself. And knowing that the results I got are not just due to a stupid mistake I made is a relief. Well there may still be other stupid mistakes in my implementation, but you get what I mean ;)
Would you mind sharing some snippets of your code? Particularly the part that creates links to other CGNS files? I'm still a beginner with CGNS, and really struggle with the documentation.

**Reply by flotus1** [reply]
@brtnfld 
> I would be curious if this would help (with CGNS develop), assuming the slowdown is in the HDF5 layer:
> 
> ```
>  INTEGER, TARGET :: value_f
>  value_f = 16*1024*1024
>  CALL cg_configure_f(CG_CONFIG_HDF5_MD_BLOCK_SIZE, C_LOC(value_f), ier)

Seems like I don't get to do that: Error: Symbol ‘cg_config_hdf5_md_block_size’ at (1) has no IMPLICIT type; did you mean ‘cg_config_hdf5_diskless_write’?
What exactly do you mean by "with cgns develop"?

**Comment by brtnfld** [comment]
You need to use the CGNS develop branch on github (it's the default when you clone it):
git clone https://github.com/CGNS/CGNS.git

Then install CGNS from the source.

**Reply by flotus1** [reply]
I see. So far I have been working with the master branch.
Gave the develop branch a quick try with your suggestion. It compiles and runs, but with the same slowdown I experienced initially.

**Reply by brtnfld** [reply]
Ok, thanks for trying.

**Comment by scspiegel** [comment]
Oh boy, I was hoping you wouldn't ask that. It's been a while since I touched that code, so I've forgotten some of the finer details on everything it does. I was able to reduce my subroutine that writes solution.time.cgns down to what I think should be a basic working example. There might be something not declared properly, not declared at all, or some general bug, but that is because I just quickly removed a bunch of extra stuff like error checking and module variables/arrays for other subroutines.

Note, this is my implementation of how to do this. There may be some easier/better ways of doing specific things, but the code seems to work so I just let it be.

```
module cgns_mod
  !
  !.. Use Statements
  use, intrinsic :: iso_c_binding, only : C_CHAR
  use cgns
  !
  implicit none
  !
  public
  !
  ! Define the base CGNS integer types
  !
  integer, parameter :: CBT = kind(0)  ! CGNS base integer type
  !
  integer, parameter :: CGLEN = 32     ! maximum length for a CGNS string
  !
  ! Last argument required for calls to cg_goto_f
  !
  character(len=*,kind=C_CHAR), parameter :: CGNS_GOTO_END = "end"
  !
  ! CGNS error variable
  !
  integer(CBT), save :: cgierr
  !
  ! ################################
  ! ###  Module Local Variables  ###
  ! ################################
  !
  ! SOLTIME stores the physical time value for each solution file
  real(r4),             save, allocatable :: soltime(:)
  ! SOLNAME stores the name of the solution node from each solution file
  character(len=CGLEN), save, allocatable :: solname(:)
  ! SOLFILE stores the actual name of each CGNS solution file
  character(len=CGLEN), save, allocatable :: solfile(:)
  !
  character(len=*), parameter :: base_name = "Base"
  character(len=*), parameter :: zone_name = "Solution"
  !
  ! ZONE_SIZE is created when writing the solution.grid.cgns file 
  integer(cgsize_t), dimension(1:3) :: zone_size
  !
contains
!
!###############################################################################
!
subroutine write_cgns_time_file()
  !
  !.. Use Statements ..
  use flow_param, only : nr  ! spatial dimension
  !
  !.. Local Scalars ..
  integer :: n
  !
  integer(CBT) :: ifile_t,ibase_t,izone_t,iiter_t
  !
  character(len=CGLEN) :: sect_name
  character(len=300)   :: link_path
  !
  !.. Local Arrays ..
  integer(cgsize_t) :: idata(1:2)
  !
  !.. Local Parameters ..
  character(len=*), parameter :: pname = "write_cgns_time_file"
  !
continue
  !
  iiter_t = 1_CBT
  !
  call cg_open_f("solution.time.cgns",CG_MODE_WRITE,ifile_t,cgierr)
  !
  ! Set the CGNS base (who knows what this is for)
  !
  call cg_base_write_f(ifile_t,base_name,int(nr,kind=CBT), &
                       int(nr,kind=CBT),ibase_t,cgierr)
  !
  ! Write the title for this solution
  !
  call cg_zone_write_f(ifile_t,ibase_t,zone_name,zone_size, &
                       UNSTRUCTURED,izone_t,cgierr)
  !
  ! Create a link to the grid coordinates in the grid file
  !
  call cg_goto_f(ifile_t,ibase_t,cgierr, &
                 "Zone_t",izone_t, &
                 CGNS_GOTO_END)
  !
  link_path = base_name//"/"//zone_name//"/GridCoordinates"
  !
  call cg_link_write_f("GridCoordinates","solution.grid.cgns", &
                       trim(link_path),cgierr)
  !
  ! Create a link to the element connectivity in the grid file
  !
  if (nr == 1) then
    sect_name = "Edges"
  else if (nr == 2) then
    sect_name = "Quadrilaterals"
  else if (nr == 3) then
    sect_name = "Hexahedrals"
  end if
  link_path = base_name//"/"//zone_name//"/"//trim(sect_name)
  !
  call cg_link_write_f(trim(sect_name),"solution.grid.cgns", &
                       trim(link_path),cgierr)
  !
  ! Create a links to the solution node in each of the solution files
  !
  do n = 1,size(solname)
    !
    link_path = base_name//"/"//zone_name//"/"//trim(solname(n))
    !
    call cg_link_write_f(trim(solname(n)),trim(solfile(n)), &
                         trim(link_path),cgierr)
    !
  end do
  !
  !
  call cg_biter_write_f(ifile_t,ibase_t,'TimeIterValues', &
                        size(soltime,kind=CBT),cgierr)
  !
  call cg_goto_f(ifile_t,ibase_t,cgierr, &
                 "BaseIterativeData_t",iiter_t, &
                 CGNS_GOTO_END)
  !
  call cg_array_write_f("TimeValues",REALSINGLE,1_CBT, &
                        size(soltime,kind=cgsize_t),soltime,cgierr)
  !
  ! Update the zone iterative node
  !
  call cg_ziter_write_f(ifile_t,ibase_t,izone_t,"ZoneIterativeData",cgierr)
  !
  call cg_goto_f(ifile_t,ibase_t,cgierr, &
                 "Zone_t",izone_t, &
                 "ZoneIterativeData_t",iiter_t, &
                 CGNS_GOTO_END)
  !
  idata(:) = int( [32,size(solname)] , kind=kind(idata) )
  call cg_array_write_f("FlowSolutionPointers",CHARACTER,2_CBT, &
                        idata,solname,cgierr)
  !
  ! Identify the flow solution within the CGNS file as time dependent
  !
  call cg_simulation_type_write_f(ifile_t,ibase_t,TIMEACCURATE,cgierr)
  !
  ! Close the CGNS FILE
  !
  call cg_close_f(ifile_t,cgierr)
  !
end subroutine write_cgns_time_file
!
!###############################################################################
!
end module cgns_mod
```

**Reply by flotus1** [reply]
You could have said no ;)
But thanks a lot for helping me out, much appreciated

**Reply by brtnfld** [reply]
If you get a stand-alone example of this to work we can add it as an example in the source for everyone to enjoy :grin:

**Comment by flotus1** [comment]
Just for the record: I consider this a workaround, not a solution to the problem.
One of the reasons why I am trying to switch to CGNS is to get away from having too many files. Some HPC clusters are limiting the number of files in addition to storage space. Having a separate file for each time step is not ideal in that regard.

**Reply by flowcntrl** [reply]
I'm having the exactly same problem... have you found any workaround other than generating a new file for every iteration?
